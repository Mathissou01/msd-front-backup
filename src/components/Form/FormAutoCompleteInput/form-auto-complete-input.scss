@use "src/styles/01-settings/settings.colors" as colors;
@use "src/styles/02-tools/tools.z-index" as index;

.c-FormAutoCompleteInput {
  display: flex;
  flex-direction: column;

  &__Container {
    position: relative;
    z-index: index.z("site-content");
    flex: 1;
  }

  &__Results {
    position: absolute;
    top: calc(100% - 4px);
    z-index: -1;
    width: 100%;
    margin: 0;
    padding: 12px 5px 8px;
    color: colors.$blue-2;
    list-style: none;
    background-color: colors.$gray-100;
    border: 1px solid colors.$blue-3;
    border-radius: 0 0 4px 4px;

    li {
      padding: 4px 10px;
      cursor: pointer;
      user-select: none;

      &:not(:last-child) {
        border-bottom: 1px dashed colors.$gray-500;
      }

      &:hover {
        background-color: colors.$gray-200;
      }
    }

    &_noResults,
    &_loading {
      color: colors.$gray-600;
      font-weight: normal;
      background-color: colors.$gray-0;

      span {
        padding: 4px 0 4px 10px;
      }
    }

    &_loading {
      &::after {
        display: inline-block;
        width: 0;
        margin-left: 2px;
        overflow: hidden;
        font-weight: 700;
        vertical-align: bottom;
        animation: ellipsis steps(4, end) 600ms infinite;
        content: "\2026"; /* ascii code for the ellipsis character */
      }

      @keyframes ellipsis {
        // stylelint-disable-next-line
        to {
          width: 1.25em;
        }
      }
    }
  }
}
