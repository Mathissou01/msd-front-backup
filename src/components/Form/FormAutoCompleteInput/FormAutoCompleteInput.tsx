import _ from "lodash";
import React, { useRef, useState } from "react";
import { ErrorMessage } from "@hookform/error-message";
import { Controller, useFormContext } from "react-hook-form";
import { removeNulls } from "../../../lib/utilities";
import CommonFormErrorText from "../../Common/CommonFormErrorText/CommonFormErrorText";
import { IFormLabelProps } from "../FormLabel/FormLabel";
import CommonInput from "../../Common/CommonInput/CommonInput";
import "./form-auto-complete-input.scss";

interface IFormAutoCompleteInputProps<T> {
  name: string;
  searchFunction: (searchValue: string) => Promise<Array<T>>;
  displayTransformFunction: (result: T) => string;
  selectTransformFunction: (result: T) => string | undefined;
  isLoading: boolean;
  isRequired?: boolean;
  minLength?: number;
  isDisabled?: boolean;
  defaultValue?: string;
  placeholder?: string;
  labelProps?: IFormLabelProps;
}

export default function FormAutoCompleteInput<T>({
  name,
  searchFunction,
  displayTransformFunction,
  selectTransformFunction,
  isLoading,
  isRequired = false,
  minLength = 3,
  isDisabled = false,
  defaultValue,
  placeholder,
}: IFormAutoCompleteInputProps<T>) {
  /* Static Data */
  const isLoadingMessage = "Recherche d'adresse";
  const noResultsMessage = "Aucune adresse correspondante à cette recherche";
  const errorMessages = {
    required: "Ce champ est obligatoire",
    minLength: `Veuillez rechercher une adresse existante en saisissant au moins ${minLength} caractères, puis sélectionnez un résultat`,
  };

  /* Methods */
  function resetResults() {
    hasResultSelected.current = false;
    if (results !== null) {
      setResults(null);
    }
  }

  function handleInput(value: string) {
    const trimmedValue = value.trim();
    if (trimmedValue && trimmedValue.length >= minLength) {
      searchFunction(trimmedValue).then((results) => {
        if (results && results.length > 0) {
          setResults(results.filter(removeNulls));
        } else {
          setResults([]);
        }
      });
    } else {
      setResults(null);
    }
  }

  function handleClickResult(result: T) {
    hasResultSelected.current = true;
    setValue(name, selectTransformFunction(result), {
      shouldDirty: true,
      shouldValidate: true,
    });
  }

  function displayResults(results: Array<T>) {
    if (results.length > 0) {
      return (
        <ul className="c-FormAutoCompleteInput__Results">
          {results.map((result, index) => {
            return (
              <li key={index} onClick={() => handleClickResult(result)}>
                {displayTransformFunction(result)}
              </li>
            );
          })}
        </ul>
      );
    } else {
      return (
        <div className="c-FormAutoCompleteInput__Results c-FormAutoCompleteInput__Results_noResults">
          <span>{noResultsMessage}</span>
        </div>
      );
    }
  }

  /* Local Data */
  const [results, setResults] = useState<Array<T> | null>(null);
  const hasResultSelected = useRef<boolean>(!!defaultValue);
  const {
    control,
    setValue,
    formState: { isSubmitting, errors },
  } = useFormContext();

  return (
    <div className="c-FormAutoCompleteInput">
      <Controller
        control={control}
        name={name}
        rules={{
          required: { value: isRequired, message: errorMessages.required },
          minLength: { value: minLength, message: errorMessages.minLength },
          validate: () => hasResultSelected.current,
        }}
        defaultValue={defaultValue}
        render={({ field: { onChange, value, ref } }) => {
          return (
            <div className="c-FormAutoCompleteInput__Container">
              <CommonInput
                id={name}
                type="text"
                value={value}
                refCallBack={ref}
                placeholder={placeholder}
                isInvalid={!!_.get(errors, name)}
                isDisabled={isDisabled || isSubmitting}
                dataTestid="form-auto-complete-input"
                onChange={(value: string) => {
                  resetResults();
                  onChange(value);
                }}
                onChangeDelayed={handleInput}
                changeDelay={350}
              />
              {!hasResultSelected.current && (
                <>
                  {isLoading && !results && (
                    <div className="c-FormAutoCompleteInput__Results c-FormAutoCompleteInput__Results_loading">
                      <span>{isLoadingMessage}</span>
                    </div>
                  )}
                  {results && displayResults(results)}
                </>
              )}
            </div>
          );
        }}
      />
      <ErrorMessage
        errors={errors}
        name={name}
        render={({ message }: { message: string }) => (
          <CommonFormErrorText message={message} errorId={`${name}_error`} />
        )}
      />
    </div>
  );
}
