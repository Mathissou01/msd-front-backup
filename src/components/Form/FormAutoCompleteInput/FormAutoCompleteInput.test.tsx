import { render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import FormAutoCompleteInput from "./FormAutoCompleteInput";

interface ITestResult {
  name: string;
}

const mock = {
  name: "autocomplete",
  label: "form autocomplete input label",
  results: [{ name: "name 1" }, { name: "name 2" }],
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>
    </>
  );
};

async function searchFunction(): Promise<Array<ITestResult>> {
  return new Promise(() => mock.results);
}

describe("FormAutoCompleteInput", () => {
  it("renders", () => {
    const { container } = render(
      <Wrapper>
        <FormAutoCompleteInput<ITestResult>
          name={mock.name}
          searchFunction={searchFunction}
          displayTransformFunction={(result) => result.name ?? ""}
          selectTransformFunction={(result) => result.name ?? undefined}
          isLoading={false}
        />
      </Wrapper>,
    );

    const input = screen.getByTestId("form-auto-complete-input");
    expect(input).toBeInTheDocument();
    expect(input).not.toHaveClass("c-CommonInput_invalid");
    expect(input).not.toHaveAttribute("disabled");

    expect(container).toMatchSnapshot();
  });
});
