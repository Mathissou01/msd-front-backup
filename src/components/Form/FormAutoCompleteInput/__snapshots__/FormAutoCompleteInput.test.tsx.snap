// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormAutoCompleteInput renders 1`] = `
<div>
  <div
    class="c-FormAutoCompleteInput"
  >
    <div
      class="c-FormAutoCompleteInput__Container"
    >
      <input
        aria-errormessage="autocomplete_error"
        aria-invalid="false"
        autocomplete="off"
        class="c-CommonInput"
        data-testid="form-auto-complete-input"
        id="autocomplete"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
